### PR TITLE
[plugin-web-app] Add step closes current window with alerts handling

### LINKED ISSUE #2314: Confirmation dialog doesn't appears when user close the current window
**Describe the bug**
"When I close the current window" step closes the current tab but doesn't catch the confirmation dialog that appears before window is closed. It's also impossible to close the tab by using Ctrl+F4 keys.

**To Reproduce**
```gherkin
Scenario: Comment is not saved if user closed tab
When I initialize the SCENARIO variable `newComment` with value `#{generate(regexify '[a-z]{20}[1-9]{5}')}`
Given I am on a page with the URL 'https://ru.wikipedia.org/wiki/%D0%A0%D1%83%D1%81%D1%81%D0%BA%D0%B0%D1%8F_%D0%92%D0%B8%D0%BA%D0%B8%D0%BF%D0%B5%D0%B4%D0%B8%D1%8F'
When I open URL `https://www.google.ru/` in new window
When I switch to window with title that contains `Русская Википедия`
When I click on element located `By.xpath(//*[@id='ca-ve-edit'])`
When I wait until element located `By.xpath(//*[@class='ve-init-mw-progressBarWidget-bar'])` disappears
When I wait until element located `By.xpath(//*[contains(@class, 'messageDialog-text')]/*[text()='Добро пожаловать в Википедию'])` appears
When I click on element located `By.xpath(//*[@role='button']/*[text()='Начать редактирование'])`
When I change context to element located `By.xpath(//*[@role='textbox'])`
When I click on element located `By.xpath(//*[@id='mwBg'])`
When I execute sequence of actions:
|type      |argument               |
|ENTER_TEXT| ${newComment}         |
When I close the current window
Then an alert is present
When I accept alert with message which contains ``
```

**Expected behavior**
"Leave this site?" dialog appears before the current tab is closed. User has possibility to accept or decline it.

**Artifacts**
2021-12-21_17h51_24

**Execution environment (please complete the following information):**
 - Vividus version: 0.3.12
 - OS: Windows 10
 - Java: jdk-11.0.12

**Desktop Web (if applicable, please complete the following information):**
 - OS: Windows 10
 - Browser chrome
 - Version 96.0.4664.110

### PR DESCRIPTION
**Pull request by draker94**
Closes #2314

### CODE REVIEW

File: docs/modules/plugins/pages/plugin-web-app.adoc [hunk 1]
```diff
@@ -225,6 +229,57 @@ When I open URL `$URL` in new window
 When I open URL `https://docs.vividus.dev` in new window
 ----
 
+=== Close current window
+
+Closes current window and switches to the previous window.
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-web-app.adoc:234]:
```suggestion
Closes the current window and switches to the previous window.
```

File: docs/modules/plugins/pages/plugin-web-app.adoc [hunk 2]
```diff
@@ -225,6 +229,57 @@ When I open URL `$URL` in new window
...
+[source,gherkin]
+----
+Given I am on a page with the URL 'https://example.com/'
+When I open URL `https://example.com/contact-us` in new window
+When I close the current window
+----
+
+=== Close current window with alert handling
+
+Trying to close current window with '{onbeforeunload-event}' events handling.
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-web-app.adoc:255]:
```suggestion
Trying to close the current window with '{onbeforeunload-event}' events handling.
```

File: docs/modules/plugins/pages/plugin-web-app.adoc [hunk 3]
```diff
@@ -225,6 +229,57 @@ When I open URL `$URL` in new window
...
+
+* If an {popup-alert}[alert] window via '{onbeforeunload-event}' event opens, it must be checked and handled in the next steps
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-web-app.adoc:257]:
```suggestion
* If an {popup-alert}[alert] window via '{onbeforeunload-event}' event is opened, it must be checked and handled in the subsequent steps.
```

File: docs/modules/plugins/pages/plugin-web-app.adoc [hunk 4]
```diff
@@ -225,6 +229,57 @@ When I open URL `$URL` in new window
...
+* If an {popup-alert}[alert] window does not open, closes current window and switches to the window from which redirected to current window
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-web-app.adoc:258]:
```suggestion
* If an {popup-alert}[alert] window is not opened, the step closes the current window and switches to the previous window.
```

File: docs/modules/plugins/pages/plugin-web-app.adoc [hunk 5]
```diff
@@ -225,6 +229,57 @@ When I open URL `$URL` in new window
...
+
+[source,gherkin]
+----
+When I try to close the current window with alerts handling
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-web-app.adoc:262]:
```suggestion
When I attempt to close current window with possibility to handle alert
```

File: docs/modules/plugins/pages/plugin-web-app.adoc [hunk 6]
```diff
@@ -225,6 +229,57 @@ When I open URL `$URL` in new window
...
+----
+
+IMPORTANT: This step can only be used if the current window was opened via the step ``When I open URL \`$pageUrl` in new window``.
+
+NOTE: If you confirm window close in {popup-alert}[alert], the window will be closed, and you will need to switch to current window using the next step: `When I switch to window with title that $stringComparisonRule `$windowName``.
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-web-app.adoc:267]:
```suggestion
NOTE: If you confirm window close in {popup-alert}[alert], the window will be closed, and you will need to switch to current window using the following step: `When I switch to window with title that $stringComparisonRule `$windowName``.
```

File: vividus-plugin-web-app/src/main/java/org/vividus/steps/ui/web/PageSteps.java [hunk 7]
```diff
@@ -281,15 +283,69 @@ public void iGoTo(String relativeURL)
      */
     @When("I close the current window")
     public void closeCurrentWindow()
+    {
+        tryToCloseCurrentWindow(false);
```
Comment (valfirst) [inline vividus-plugin-web-app/src/main/java/org/vividus/steps/ui/web/PageSteps.java:287]: could you please move this step and next one to `WindowSteps` class

File: vividus-plugin-web-app/src/main/java/org/vividus/steps/ui/web/PageSteps.java [hunk 8]
```diff
@@ -281,15 +283,69 @@ public void iGoTo(String relativeURL)
...
+     * @see <a href="https://www.w3schools.com/jsref/event_onbeforeunload.asp"><i>Event 'onbeforeunload'</i></a>
+     * @see <a href="https://www.w3schools.com/js/js_popup.asp"><i>JavaScript Popup Boxes</i></a>
+     */
+    @When("I try to close the current window with alerts handling")
+    public void closeCurrentWindowWithAlertsHandling()
+    {
+        tryToCloseCurrentWindow(true);
+    }
+
+    private void tryToCloseCurrentWindow(boolean isAlertHandling)
```
Comment (valfirst) [inline vividus-plugin-web-app/src/main/java/org/vividus/steps/ui/web/PageSteps.java:324]:
```suggestion
    private void tryToCloseCurrentWindow(boolean hadnleAlerts)
```

File: vividus-plugin-web-app/src/main/java/org/vividus/steps/ui/web/PageSteps.java [hunk 9]
```diff
@@ -281,15 +283,69 @@ public void iGoTo(String relativeURL)
...
+                if (isAlertHandling)
+                {
+                    javascriptActions.closeCurrentWindow();
+                    if (!alertActions.isAlertPresent())
+                    {
+                        driver.switchTo().window(window);
+                    }
+                    else
+                    {
+                        LOGGER.info("Alert dialog box appears and must be processed.");
```
Comment (valfirst) [inline vividus-plugin-web-app/src/main/java/org/vividus/steps/ui/web/PageSteps.java:341]:
```suggestion
                        LOGGER.info("Alert dialog box is shown and must be handled in the subsequent steps.");
```

File: vividus-plugin-web-app/src/main/java/org/vividus/ui/web/action/WebJavascriptActions.java [hunk 10]
```diff
@@ -149,6 +149,11 @@ public void openNewWindow()
         executeScript("window.open()");
     }
 
+    public void closeCurrentWindow()
+    {
+        executeScript("window.close()");
```
Comment (valfirst) [inline vividus-plugin-web-app/src/main/java/org/vividus/ui/web/action/WebJavascriptActions.java:154]: need to be covered by unit tests

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-plugin-web-app/src/main/java/org/vividus/steps/ui/web/PageSteps.java [hunk 11]
```diff
@@ -281,15 +283,69 @@ public void iGoTo(String relativeURL)
...
     {
         WebDriver driver = getWebDriver();
         String currentWindow = driver.getWindowHandle();
         for (String window : driver.getWindowHandles())
         {
             if (!window.equals(currentWindow))
             {
-                driver.close();
-                driver.switchTo().window(window);
+                if (isAlertHandling)
```
Comment (ikalinin1) [inline vividus-plugin-web-app/src/main/java/org/vividus/steps/ui/web/PageSteps.java:332]: I propose to separate methods with/without alert handling and remove this condition. We always know if alert handling is required, so why not use the proper method

Comment (draker94) [reply vividus-plugin-web-app/src/main/java/org/vividus/steps/ui/web/PageSteps.java:332]: > I propose to separate methods with/without alert handling and remove this condition. We always know if alert handling is required, so why not use the proper method

The main goal of creating the common method was to get rid of code duplication. If we will put current common logic in different methods, we will have to duplicate the code or move the common logic into several more methods. Or I misunderstand you?

File: vividus-tests/src/main/resources/story/integration/SetContextStepsTests.story [hunk 12]
```diff
@@ -86,3 +86,28 @@ Given I am on a page with the URL '${vividus-test-site-url}'
 When I change context to element located `tagName(a)`
 When I execute javascript `location.reload();` with arguments:
 Then number of elements found by `cssSelector(img)` is = `1`
+
+
+Scenario: Verify step: "When I attempt to close current window with possibility to handle alert" with alert
+Given I am on a page with the URL '${vividus-test-site-url}'
```
Comment (valfirst) [inline vividus-tests/src/main/resources/story/integration/SetContextStepsTests.story:92]: is it needed?

File: vividus-tests/src/main/resources/story/integration/SetContextStepsTests.story [hunk 13]
```diff
@@ -86,3 +86,28 @@ Given I am on a page with the URL '${vividus-test-site-url}'
...
+When I open URL `${vividus-test-site-url}/onbeforeunloadAlert.html` in new window
+Then an alert is not present
+When I click on element located `By.xpath(//a[@href='javascript:void()'])`
```
Comment (valfirst) [inline vividus-tests/src/main/resources/story/integration/SetContextStepsTests.story:95]:
```suggestion
When I click on element located `By.xpath(//a[text() = 'here'])`
```

File: vividus-plugin-web-app/src/main/java/org/vividus/steps/ui/web/WindowSteps.java [hunk 14]
```diff
@@ -16,25 +16,42 @@
...
+
     private final IWebDriverProvider webDriverProvider;
     private final IWebDriverManager webDriverManager;
     private final ISoftAssert softAssert;
+    private final WebJavascriptActions javascriptActions;
+    private final IAlertActions alertActions;
 
-    public WindowSteps(IWebDriverProvider webDriverProvider, IWebDriverManager webDriverManager, ISoftAssert softAssert)
+    public WindowSteps(IWebDriverProvider webDriverProvider, IWebDriverManager webDriverManager, ISoftAssert softAssert,
+            WebJavascriptActions javascriptActions, IAlertActions alertActions)
```
Comment (valfirst) [inline vividus-plugin-web-app/src/main/java/org/vividus/steps/ui/web/WindowSteps.java:48]:
```suggestion
    public WindowSteps(IWebDriverProvider webDriverProvider, IWebDriverManager webDriverManager, WebJavascriptActions javascriptActions, IAlertActions alertActions, ISoftAssert softAssert)
```

Comment (valfirst) [reply vividus-plugin-web-app/src/main/java/org/vividus/steps/ui/web/WindowSteps.java:48]: the same order should be applied to fields and the order of fields initialization in the  conctructor

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-plugin-web-app/src/main/java/org/vividus/steps/ui/web/WindowSteps.java [hunk 15]
```diff
@@ -49,10 +66,96 @@ public void resizeCurrentWindow(Dimension targetSize)
...
+        tryToCloseCurrentWindow((driver, window) ->
+        {
+            driver.close();
+            driver.switchTo().window(window);
+        });
+    }
+
+    /**
+     * Trying to close the <b>current window</b> with JavaScript method 'close()'.
+     * If an alert window via 'onbeforeunload' event is opened, it must be checked and handled in the subsequent steps.
```
Comment (uarlouski) [inline vividus-plugin-web-app/src/main/java/org/vividus/steps/ui/web/WindowSteps.java:101]: If an alert window is opened via 'onbeforeunload' ...
imho it's more clear to read

File: vividus-plugin-web-app/src/main/java/org/vividus/steps/ui/web/WindowSteps.java [hunk 16]
```diff
@@ -49,10 +66,96 @@ public void resizeCurrentWindow(Dimension targetSize)
     {
         if (canTryToResize(targetSize))
         {
-            webDriverProvider.get().manage().window().setSize(targetSize);
+            getWebDriver().manage().window().setSize(targetSize);
         }
     }
 
+    /**
+     * Closes <b>current window</b> and switches to the window from which rederection to current window was performed
```
Comment (uarlouski) [inline vividus-plugin-web-app/src/main/java/org/vividus/steps/ui/web/WindowSteps.java:74]: rederiction

**Review by ikalinin1 — APPROVED** [review]

**Review by valfirst — APPROVED** [review]

**Review by uarlouski — APPROVED** [review]

**Pull request merged**